compat/src/main/scala-2.11_2.12/scala/collection/compat/CompatImpl.scala

```diff
@@ -12,10 +12,47 @@
 
 package scala.collection.compat
 
+import scala.reflect.ClassTag
 import scala.collection.generic.CanBuildFrom
 import scala.collection.mutable.Builder
 import scala.collection.{immutable => i, mutable => m}
 
+/* builder optimized for a single ++= call, which returns identity on result if possible
+ * and defers to the underlying builder if not.
+ */
+private final class IdentityPreservingBuilder[A, CC[X] <: TraversableOnce[X]](that: Builder[A, CC[A]])(implicit ct: ClassTag[CC[A]])
+    extends Builder[A, CC[A]] {
+  var collection: CC[A] = null.asInstanceOf[CC[A]]
+  var ruined = false
+
+  final override def ++=(elems: TraversableOnce[A]): this.type =
+      elems match {
+        case ct(ca) if (collection == null && !ruined) =>  {
+          collection = ca
+          this
+        }
+        case _ => {
+          ruined = true
+          if (collection != null) that ++= collection
+          that ++= elems
+          collection = null.asInstanceOf[CC[A]]
+          this
+        }
+      }
+
+  final def +=(elem: A): this.type = {
+    collection = null.asInstanceOf[CC[A]]
```

Comment:
does this need to do `if (collection != null) that ++= collection`? should that logic be extracted out into a(n inline?) method?

```scala
private def ruin(): Unit = {
  ruined = true
  val coll = collection
  if (coll != null) {
    that ++= coll
    collection = null.asInstanceOf[CC[A]]
  }
}
```

Reply:
I did a general update of ruination, which is, I think, a bit tighter still.